controller/cmd/multus/main.go

```diff
@@ -0,0 +1,100 @@
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	multusapi "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+	multusctrl "github.com/linkerd/linkerd2/controller/multus"
+)
+
+// Main executes the identity subcommand
+func Main(args []string) {
+	cmd := flag.NewFlagSet("multus", flag.ExitOnError)
+
+	metricsAddr := cmd.String("metrics-address", ":8081", "Prometheus metrics bind address")
+	probeAddr := cmd.String("probe-address", ":8082", "Health probe address")
+	enableLeaderElection := cmd.Bool("leader-election", true, "Enable controller leader election")
+	cniNamespace := cmd.String("cni-namespace", "linkerd-cni", "namespace in which Linkerd-CNI is installed")
+	cniKubeconfigFilePath := cmd.String("cni-kubeconfig", "/etc/cni/net.d/ZZZ-linkerd-cni-kubeconfig", "Linkerd-CNI Kubeconfig path")
+	linkerdNamespace := cmd.String("linkerd-namespace", "", "namespace in which Linkerd is installed")
+
+	componentName := "linkerd-multus.linkerd.io"
+
+	opts := zap.Options{
+		Development: true,
+	}
+	opts.BindFlags(cmd)
+
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	flags.ConfigureAndParse(cmd, args)
+
+	setupLogger := ctrl.Log.WithName("setup")
+
+	setupLogger.Info("Starting controller with settings",
+		"metricsAddr", *metricsAddr,
+		"healthProbeAddr", *probeAddr,
+		"enableLeaderSelectiion", *enableLeaderElection,
+		"cniNamespace", *cniNamespace,
+		"cniKubeconfigPath", *cniKubeconfigFilePath,
+		"componentName", componentName,
+		"linkerdNamespace", *linkerdNamespace,
+	)
+
+	// Manager.
+	var scheme = runtime.NewScheme()
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(multusapi.AddToScheme(scheme))
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:                 scheme,
+		MetricsBindAddress:     *metricsAddr,
+		HealthProbeBindAddress: *probeAddr,
+		LeaderElection:         *enableLeaderElection,
+		LeaderElectionID:       componentName,
+	})
+	if err != nil {
+		setupLogger.Error(err, "Failed to setup Manager")
+		os.Exit(1)
+	}
+
+	//
+	// Create, initialize and run service.
+	//
+	cniGen := multusctrl.NewAttachReconciler(context.Background(), mgr.GetClient(),
+		componentName, *cniNamespace, *cniKubeconfigFilePath, *linkerdNamespace)
+
+	if err := cniGen.SetupWithManager(mgr); err != nil {
+		setupLogger.Error(err, "Can not setup controller")
+		os.Exit(1)
+	}
+
+	// Probes.
+	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+		setupLogger.Error(err, "unable to set up health check")
+		os.Exit(1)
+	}
+	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+		setupLogger.Error(err, "unable to set up ready check")
+		os.Exit(1)
+	}
+
+	// Start.
+	if err := mgr.Start(context.Background()); err != nil {
```

Comment:
For better or worse, Linkerd doesn't use controller-runtime anywhere, so this controller looks different from the others in the core control plane. I have a strong preference to keep our components relatively uniform.

I also appreciate that this controller necessarily behaves differently than our other components (it's the only one that is primarily responsible for maintaining resources in the cluster), so it may be desirable to use the framework for this use case.